### ISSUE TITLE
pipgrep allennlp is slow

### ISSUE THREAD

**Issue by tekumara** [ddelange/pipgrip#110]
#### What you were trying to do (and why)

`pipgrep allennlp` takes > 10 mins.

#### What happened (including command output)

<details><summary>Command output</summary><p>

```console
$ pipgrip -vvv <your-command>
❯ pipgrip -vv allennlp
INFO: fact: _root_ is root
INFO: derived: root
INFO: fact: root depends on allennlp (*)
INFO: selecting _root_ (0.0.0)
INFO: derived: allennlp (*)
INFO: discovering allennlp
INFO: fact: allennlp (2.10.1) depends on base58 (>=2.1.1)
INFO: fact: allennlp (2.10.1) depends on cached-path (>=1.1.3,<1.2.0)
INFO: fact: allennlp (2.10.1) depends on dill (>=0.3.4)
INFO: fact: allennlp (2.10.1) depends on fairscale (==0.4.6)
INFO: fact: allennlp (2.10.1) depends on filelock (>=3.3,<3.8)
INFO: fact: allennlp (2.10.1) depends on h5py (>=3.6.0)
INFO: fact: allennlp (2.10.1) depends on huggingface-hub (>=0.0.16)
INFO: fact: allennlp (2.10.1) depends on jsonnet (>=0.10.0)
INFO: fact: allennlp (2.10.1) depends on lmdb (>=1.2.1)
INFO: fact: allennlp (2.10.1) depends on more-itertools (>=8.12.0)
INFO: fact: allennlp (2.10.1) depends on nltk (>=3.6.5)
INFO: fact: allennlp (2.10.1) depends on numpy (>=1.21.4)
INFO: fact: allennlp (2.10.1) depends on protobuf (<4.0.0,>=3.12.0)
INFO: fact: allennlp (2.10.1) depends on pytest (>=6.2.5)
INFO: fact: allennlp (2.10.1) depends on requests (>=2.28)
INFO: fact: allennlp (2.10.1) depends on sacremoses (*)
INFO: fact: allennlp (2.10.1) depends on scikit-learn (>=1.0.1)
INFO: fact: allennlp (2.10.1) depends on scipy (>=1.7.3)
INFO: fact: allennlp (2.10.1) depends on sentencepiece (>=0.1.96)
INFO: fact: allennlp (2.10.1) depends on spacy (>=2.1.0,<3.4)
INFO: fact: allennlp (2.10.1) depends on tensorboardx (>=1.2)
INFO: fact: allennlp (2.10.1) depends on termcolor (==1.1.0)
INFO: fact: allennlp (2.10.1) depends on torch (<1.13.0,>=1.10.0)
INFO: fact: allennlp (2.10.1) depends on torchvision (<0.14.0,>=0.8.1)
INFO: fact: allennlp (2.10.1) depends on tqdm (>=4.62)
INFO: fact: allennlp (2.10.1) depends on traitlets (>5.1.1)
INFO: fact: allennlp (2.10.1) depends on transformers (<4.21,>=4.1)
INFO: fact: allennlp (2.10.1) depends on typer (>=0.4.1)
INFO: fact: allennlp (2.10.1) depends on wandb (>=0.10.0,<0.13.0)
INFO: selecting allennlp (2.10.1)
INFO: derived: wandb (>=0.10.0,<0.13.0)
INFO: derived: typer (>=0.4.1)
INFO: derived: transformers (<4.21,>=4.1)
INFO: derived: traitlets (>5.1.1)
INFO: derived: tqdm (>=4.62)
INFO: derived: torchvision (<0.14.0,>=0.8.1)
INFO: derived: torch (<1.13.0,>=1.10.0)
INFO: derived: termcolor (==1.1.0)
INFO: derived: tensorboardx (>=1.2)
INFO: derived: spacy (>=2.1.0,<3.4)
INFO: derived: sentencepiece (>=0.1.96)
INFO: derived: scipy (>=1.7.3)
INFO: derived: scikit-learn (>=1.0.1)
INFO: derived: sacremoses (*)
INFO: derived: requests (>=2.28)
INFO: derived: pytest (>=6.2.5)
INFO: derived: protobuf (<4.0.0,>=3.12.0)
INFO: derived: numpy (>=1.21.4)
INFO: derived: nltk (>=3.6.5)
INFO: derived: more-itertools (>=8.12.0)
INFO: derived: lmdb (>=1.2.1)
INFO: derived: jsonnet (>=0.10.0)
INFO: derived: huggingface-hub (>=0.0.16)
INFO: derived: h5py (>=3.6.0)
INFO: derived: filelock (>=3.3,<3.8)
INFO: derived: fairscale (==0.4.6)
INFO: derived: dill (>=0.3.4)
INFO: derived: cached-path (>=1.1.3,<1.2.0)
INFO: derived: base58 (>=2.1.1)
INFO: discovering wandb<0.13.0,>=0.10.0
INFO: discovering typer>=0.4.1
INFO: discovering transformers<4.21,>=4.1
INFO: discovering traitlets>5.1.1
INFO: discovering tqdm>=4.62
INFO: discovering torchvision<0.14.0,>=0.8.1
INFO: discovering torch<1.13.0,>=1.10.0
INFO: discovering termcolor==1.1.0
INFO: discovering tensorboardx>=1.2
INFO: discovering spacy<3.4,>=2.1.0
INFO: discovering sentencepiece>=0.1.96
INFO: discovering scipy>=1.7.3
INFO: discovering scikit-learn>=1.0.1
INFO: discovering sacremoses
INFO: discovering requests>=2.28
INFO: discovering pytest>=6.2.5
INFO: discovering protobuf<4.0.0,>=3.12.0
INFO: discovering numpy>=1.21.4
INFO: discovering nltk>=3.6.5
INFO: discovering more-itertools>=8.12.0
INFO: discovering lmdb>=1.2.1
INFO: discovering jsonnet>=0.10.0
INFO: discovering huggingface-hub>=0.0.16
INFO: discovering h5py>=3.6.0
INFO: discovering filelock<3.8,>=3.3
INFO: discovering fairscale==0.4.6
INFO: discovering dill>=0.3.4
INFO: discovering cached-path<1.2.0,>=1.1.3
INFO: discovering base58>=2.1.1
INFO: selecting termcolor (1.1.0)
INFO: selecting base58 (2.1.1)
INFO: fact: fairscale (0.4.6) depends on torch (>=1.8.0)
INFO: selecting fairscale (0.4.6)
INFO: selecting sentencepiece (0.1.98)
INFO: fact: h5py (3.8.0) depends on numpy (>=1.14.5)
INFO: selecting h5py (3.8.0)
INFO: selecting lmdb (1.4.1)
INFO: selecting dill (0.3.6)
INFO: fact: requests (2.29.0) depends on certifi (>=2017.4.17)
INFO: fact: requests (2.29.0) depends on charset-normalizer (<4,>=2)
INFO: fact: requests (2.29.0) depends on idna (<4,>=2.5)
INFO: fact: requests (2.29.0) depends on urllib3 (>=1.21.1,<1.27)
INFO: selecting requests (2.29.0)
INFO: derived: urllib3 (>=1.21.1,<1.27)
INFO: derived: idna (<4,>=2.5)
INFO: derived: charset-normalizer (<4,>=2)
INFO: derived: certifi (>=2017.4.17)
INFO: discovering urllib3<1.27,>=1.21.1
INFO: discovering idna<4,>=2.5
INFO: discovering charset-normalizer<4,>=2
INFO: discovering certifi>=2017.4.17
INFO: fact: cached-path (1.1.6) depends on boto3 (>=1.0,<2.0)
INFO: fact: cached-path (1.1.6) depends on filelock (>=3.4,<3.9)
INFO: fact: cached-path (1.1.6) depends on google-cloud-storage (<3.0,>=1.32.0)
INFO: fact: cached-path (1.1.6) depends on huggingface-hub (<0.11.0,>=0.8.1)
INFO: fact: cached-path (1.1.6) depends on requests (<3.0,>=2.0)
INFO: fact: cached-path (1.1.6) depends on rich (<13.0,>=12.1)
INFO: selecting cached-path (1.1.6)
INFO: derived: rich (<13.0,>=12.1)
INFO: derived: huggingface-hub (<0.11.0,>=0.8.1)
INFO: derived: google-cloud-storage (<3.0,>=1.32.0)
INFO: derived: filelock (>=3.4,<3.9)
INFO: derived: boto3 (>=1.0,<2.0)
INFO: discovering huggingface-hub==0.10.1
INFO: discovering rich<13.0,>=12.1
INFO: discovering google-cloud-storage<3.0,>=1.32.0
INFO: discovering boto3<2.0,>=1.0
INFO: selecting more-itertools (9.1.0)
INFO: fact: huggingface-hub (0.10.1) depends on filelock (*)
INFO: fact: huggingface-hub (0.10.1) depends on packaging (>=20.9)
INFO: fact: huggingface-hub (0.10.1) depends on pyyaml (>=5.1)
INFO: fact: huggingface-hub (0.10.1) depends on requests (*)
INFO: fact: huggingface-hub (0.10.1) depends on tqdm (*)
INFO: fact: huggingface-hub (0.10.1) depends on typing-extensions (>=3.7.4.3)
INFO: selecting huggingface-hub (0.10.1)
INFO: derived: typing-extensions (>=3.7.4.3)
INFO: derived: pyyaml (>=5.1)
INFO: derived: packaging (>=20.9)
INFO: discovering typing-extensions>=3.7.4.3
INFO: discovering pyyaml>=5.1
INFO: discovering packaging>=20.9
INFO: fact: typer (0.7.0) depends on click (<9.0.0,>=7.1.1)
INFO: selecting typer (0.7.0)
INFO: derived: click (<9.0.0,>=7.1.1)
INFO: discovering click<9.0.0,>=7.1.1
INFO: fact: torch (1.12.1) depends on typing-extensions (*)
INFO: selecting torch (1.12.1)
INFO: fact: nltk (3.8.1) depends on click (*)
INFO: fact: nltk (3.8.1) depends on joblib (*)
INFO: fact: nltk (3.8.1) depends on regex (>=2021.8.3)
INFO: fact: nltk (3.8.1) depends on tqdm (*)
INFO: selecting nltk (3.8.1)
INFO: derived: regex (>=2021.8.3)
INFO: derived: joblib (*)
INFO: discovering regex>=2021.8.3
INFO: discovering joblib
INFO: fact: torchvision (0.13.1) depends on numpy (*)
INFO: fact: torchvision (0.13.1) depends on pillow (>=5.3.0,<8.3.0 || >=8.4.0)
INFO: fact: torchvision (0.13.1) depends on requests (*)
INFO: fact: torchvision (0.13.1) depends on torch (*)
INFO: fact: torchvision (0.13.1) depends on typing-extensions (*)
INFO: selecting torchvision (0.13.1)
INFO: derived: pillow (>=5.3.0,<8.3.0 || >=8.4.0)
INFO: discovering pillow!=8.3.*,>=5.3.0
INFO: selecting filelock (3.7.1)
INFO: selecting packaging (23.1)
INFO: selecting pyyaml (6.0)
INFO: fact: scipy (1.10.1) depends on numpy (<1.27.0,>=1.19.5)
INFO: selecting scipy (1.10.1)
INFO: derived: numpy (<1.27.0,>=1.19.5)
INFO: fact: scikit-learn (1.2.2) depends on joblib (>=1.1.1)
INFO: fact: scikit-learn (1.2.2) depends on numpy (>=1.17.3)
INFO: fact: scikit-learn (1.2.2) depends on scipy (>=1.3.2)
INFO: fact: scikit-learn (1.2.2) depends on threadpoolctl (>=2.0.0)
INFO: selecting scikit-learn (1.2.2)
INFO: derived: threadpoolctl (>=2.0.0)
INFO: derived: joblib (>=1.1.1)
INFO: discovering threadpoolctl>=2.0.0
INFO: selecting joblib (1.2.0)
INFO: selecting threadpoolctl (3.1.0)
INFO: selecting tqdm (4.65.0)
INFO: selecting idna (3.4)
INFO: selecting click (8.1.3)
INFO: fact: rich (12.6.0) depends on commonmark (>=0.9.0,<0.10.0)
INFO: fact: rich (12.6.0) depends on pygments (<3.0.0,>=2.6.0)
INFO: selecting rich (12.6.0)
INFO: derived: pygments (<3.0.0,>=2.6.0)
INFO: derived: commonmark (>=0.9.0,<0.10.0)
INFO: discovering pygments<3.0.0,>=2.6.0
INFO: discovering commonmark<0.10.0,>=0.9.0
INFO: selecting commonmark (0.9.1)
INFO: selecting traitlets (5.9.0)
INFO: selecting typing-extensions (4.5.0)
INFO: fact: pytest (7.3.1) depends on exceptiongroup (>=1.0.0rc8)
INFO: fact: pytest (7.3.1) depends on iniconfig (*)
INFO: fact: pytest (7.3.1) depends on packaging (*)
INFO: fact: pytest (7.3.1) depends on pluggy (<2.0,>=0.12)
INFO: fact: pytest (7.3.1) depends on tomli (>=1.0.0)
INFO: selecting pytest (7.3.1)
INFO: derived: tomli (>=1.0.0)
INFO: derived: pluggy (<2.0,>=0.12)
INFO: derived: iniconfig (*)
INFO: derived: exceptiongroup (>=1.0.0rc8)
INFO: discovering tomli>=1.0.0
INFO: discovering pluggy<2.0,>=0.12
INFO: discovering iniconfig
INFO: discovering exceptiongroup>=1.0.0rc8
INFO: selecting pluggy (1.0.0)
INFO: selecting iniconfig (2.0.0)
INFO: selecting exceptiongroup (1.1.1)
INFO: selecting tomli (2.0.1)
INFO: selecting jsonnet (0.20.0)
INFO: fact: tensorboardx (2.6) depends on numpy (*)
INFO: fact: tensorboardx (2.6) depends on packaging (*)
INFO: fact: tensorboardx (2.6) depends on protobuf (<4,>=3.8.0)
INFO: selecting tensorboardx (2.6)
INFO: selecting numpy (1.24.3)
INFO: selecting charset-normalizer (3.1.0)
INFO: selecting pygments (2.15.1)
INFO: selecting regex (2023.3.23)
INFO: selecting pillow (9.5.0)
INFO: selecting certifi (2022.12.7)
INFO: fact: google-cloud-storage (2.8.0) depends on google-api-core (>=1.31.5,<2.0.0 || >2.3.0,<3.0.0dev)
INFO: fact: google-cloud-storage (2.8.0) depends on google-auth (<3.0dev,>=1.25.0)
INFO: fact: google-cloud-storage (2.8.0) depends on google-cloud-core (<3.0dev,>=2.3.0)
INFO: fact: google-cloud-storage (2.8.0) depends on google-resumable-media (>=2.3.2)
INFO: fact: google-cloud-storage (2.8.0) depends on requests (<3.0.0dev,>=2.18.0)
INFO: selecting google-cloud-storage (2.8.0)
INFO: derived: google-resumable-media (>=2.3.2)
INFO: derived: google-cloud-core (<3.0dev,>=2.3.0)
INFO: derived: google-auth (<3.0dev,>=1.25.0)
INFO: derived: google-api-core (>=1.31.5,<2.0.0 || >2.3.0,<3.0.0dev)
INFO: discovering google-resumable-media>=2.3.2
INFO: discovering google-cloud-core<3.0dev,>=2.3.0
INFO: discovering google-auth<3.0dev,>=1.25.0
INFO: discovering google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5
INFO: fact: google-cloud-core (2.3.2) depends on google-api-core (>=1.31.6,<2.0.0 || >2.3.0,<3.0.0dev)
INFO: fact: google-cloud-core (2.3.2) depends on google-auth (<3.0dev,>=1.25.0)
INFO: selecting google-cloud-core (2.3.2)
INFO: derived: google-api-core (>=1.31.6,<2.0.0 || >2.3.0,<3.0.0dev)
INFO: fact: google-resumable-media (2.5.0) depends on google-crc32c (>=1.0,<2.0dev)
INFO: selecting google-resumable-media (2.5.0)
INFO: derived: google-crc32c (>=1.0,<2.0dev)
INFO: discovering google-crc32c<2.0dev,>=1.0
INFO: selecting google-crc32c (1.5.0)
INFO: fact: google-api-core (2.11.0) depends on google-auth (<3.0dev,>=2.14.1)
INFO: fact: google-api-core (2.11.0) depends on googleapis-common-protos (<2.0dev,>=1.56.2)
INFO: fact: google-api-core (2.11.0) depends on protobuf (>=3.19.5,<3.20.0 || >3.20.0,<3.20.1 || >3.20.1,<4.21.0 || >4.21.0,<4.21.1 || >4.21.1,<4.21.2 || >4.21.2,<4.21.3 || >4.21.3,<4.21.4 || >4.21.4,<4.21.5 || >4.21.5,<5.0.0dev)
INFO: fact: google-api-core (2.11.0) depends on requests (<3.0.0dev,>=2.18.0)
INFO: selecting google-api-core (2.11.0)
INFO: derived: protobuf (>=3.19.5,<3.20.0 || >3.20.0,<3.20.1 || >3.20.1,<4.21.0 || >4.21.0,<4.21.1 || >4.21.1,<4.21.2 || >4.21.2,<4.21.3 || >4.21.3,<4.21.4 || >4.21.4,<4.21.5 || >4.21.5,<5.0.0dev)
INFO: derived: googleapis-common-protos (<2.0dev,>=1.56.2)
INFO: derived: google-auth (<3.0dev,>=2.14.1)
INFO: discovering googleapis-common-protos<2.0dev,>=1.56.2
INFO: selecting protobuf (3.20.3)
INFO: fact: googleapis-common-protos (1.59.0) depends on protobuf (>=3.19.5,<3.20.0 || >3.20.0,<3.20.1 || >3.20.1,<4.21.1 || >4.21.1,<4.21.2 || >4.21.2,<4.21.3 || >4.21.3,<4.21.4 || >4.21.4,<4.21.5 || >4.21.5,<5.0.0dev)
INFO: selecting googleapis-common-protos (1.59.0)
INFO: fact: google-auth (2.17.3) depends on cachetools (>=2.0.0,<6.0)
INFO: fact: google-auth (2.17.3) depends on pyasn1-modules (>=0.2.1)
INFO: fact: google-auth (2.17.3) depends on rsa (<5,>=3.1.4)
INFO: fact: google-auth (2.17.3) depends on six (>=1.9.0)
INFO: selecting google-auth (2.17.3)
INFO: derived: six (>=1.9.0)
INFO: derived: rsa (<5,>=3.1.4)
INFO: derived: pyasn1-modules (>=0.2.1)
INFO: derived: cachetools (>=2.0.0,<6.0)
INFO: discovering six>=1.9.0
INFO: discovering rsa<5,>=3.1.4
INFO: discovering pyasn1-modules>=0.2.1
INFO: discovering cachetools<6.0,>=2.0.0
INFO: selecting six (1.16.0)
INFO: fact: pyasn1-modules (0.3.0) depends on pyasn1 (>=0.4.6,<0.6.0)
INFO: selecting pyasn1-modules (0.3.0)
INFO: derived: pyasn1 (>=0.4.6,<0.6.0)
INFO: discovering pyasn1<0.6.0,>=0.4.6
INFO: selecting pyasn1 (0.5.0)
INFO: selecting cachetools (5.3.0)
INFO: fact: rsa (4.9) depends on pyasn1 (>=0.1.3)
INFO: selecting rsa (4.9)
INFO: selecting urllib3 (1.26.15)
INFO: fact: sacremoses (0.0.53) depends on click (*)
INFO: fact: sacremoses (0.0.53) depends on joblib (*)
INFO: fact: sacremoses (0.0.53) depends on regex (*)
INFO: fact: sacremoses (0.0.53) depends on six (*)
INFO: fact: sacremoses (0.0.53) depends on tqdm (*)
INFO: selecting sacremoses (0.0.53)
INFO: fact: spacy (3.3.2) depends on blis (<0.8.0,>=0.4.0)
INFO: fact: spacy (3.3.2) depends on catalogue (>=2.0.6,<2.1.0)
INFO: fact: spacy (3.3.2) depends on cymem (>=2.0.2,<2.1.0)
INFO: fact: spacy (3.3.2) depends on jinja2 (*)
INFO: fact: spacy (3.3.2) depends on langcodes (<4.0.0,>=3.2.0)
INFO: fact: spacy (3.3.2) depends on murmurhash (<1.1.0,>=0.28.0)
INFO: fact: spacy (3.3.2) depends on numpy (>=1.15.0)
INFO: fact: spacy (3.3.2) depends on packaging (>=20.0)
INFO: fact: spacy (3.3.2) depends on pathy (>=0.3.5)
INFO: fact: spacy (3.3.2) depends on preshed (<3.1.0,>=3.0.2)
INFO: fact: spacy (3.3.2) depends on pydantic (>=1.7.4,<1.8 || >1.8,<1.8.1 || >1.8.1,<1.9.0)
INFO: fact: spacy (3.3.2) depends on requests (<3.0.0,>=2.13.0)
INFO: fact: spacy (3.3.2) depends on setuptools (*)
INFO: fact: spacy (3.3.2) depends on smart-open (<7.0.0,>=5.2.1)
INFO: fact: spacy (3.3.2) depends on spacy-legacy (<3.1.0,>=3.0.9)
INFO: fact: spacy (3.3.2) depends on spacy-loggers (>=1.0.0,<2.0.0)
INFO: fact: spacy (3.3.2) depends on srsly (<3.0.0,>=2.4.3)
INFO: fact: spacy (3.3.2) depends on thinc (>=8.0.14,<8.1.0)
INFO: fact: spacy (3.3.2) depends on tqdm (<5.0.0,>=4.38.0)
INFO: fact: spacy (3.3.2) depends on typer (<0.5.0,>=0.3.0)
INFO: fact: spacy (3.3.2) depends on wasabi (<1.1.0,>=0.9.1)
INFO: derived: not spacy (3.3.2)
INFO: discovering spacy==3.3.1
INFO: fact: spacy (3.3.1) depends on blis (<0.8.0,>=0.4.0)
INFO: fact: spacy (3.3.1) depends on catalogue (>=2.0.6,<2.1.0)
INFO: fact: spacy (3.3.1) depends on cymem (>=2.0.2,<2.1.0)
INFO: fact: spacy (3.3.1) depends on jinja2 (*)
INFO: fact: spacy (3.3.1) depends on langcodes (<4.0.0,>=3.2.0)
INFO: fact: spacy (3.3.1) depends on murmurhash (<1.1.0,>=0.28.0)
INFO: fact: spacy (3.3.1) depends on numpy (>=1.15.0)
INFO: fact: spacy (3.3.1) depends on packaging (>=20.0)
INFO: fact: spacy (3.3.1) depends on pathy (>=0.3.5)
INFO: fact: spacy (3.3.1) depends on preshed (<3.1.0,>=3.0.2)
INFO: fact: spacy (3.3.1) depends on pydantic (>=1.7.4,<1.8 || >1.8,<1.8.1 || >1.8.1,<1.9.0)
INFO: fact: spacy (3.3.1) depends on requests (<3.0.0,>=2.13.0)
INFO: fact: spacy (3.3.1) depends on setuptools (*)
INFO: fact: spacy (3.3.1) depends on spacy-legacy (<3.1.0,>=3.0.9)
INFO: fact: spacy (3.3.1) depends on spacy-loggers (>=1.0.0,<2.0.0)
INFO: fact: spacy (3.3.1) depends on srsly (<3.0.0,>=2.4.3)
INFO: fact: spacy (3.3.1) depends on thinc (>=8.0.14,<8.1.0)
INFO: fact: spacy (3.3.1) depends on tqdm (<5.0.0,>=4.38.0)
INFO: fact: spacy (3.3.1) depends on typer (<0.5.0,>=0.3.0)
INFO: fact: spacy (3.3.1) depends on wasabi (<1.1.0,>=0.9.1)
INFO: derived: not spacy (3.3.1)
INFO: discovering spacy==3.3.0
INFO: fact: spacy (3.3.0) depends on blis (<0.8.0,>=0.4.0)
INFO: fact: spacy (3.3.0) depends on catalogue (>=2.0.6,<2.1.0)
INFO: fact: spacy (3.3.0) depends on cymem (>=2.0.2,<2.1.0)
INFO: fact: spacy (3.3.0) depends on jinja2 (*)
INFO: fact: spacy (3.3.0) depends on langcodes (<4.0.0,>=3.2.0)
INFO: fact: spacy (3.3.0) depends on murmurhash (<1.1.0,>=0.28.0)
INFO: fact: spacy (3.3.0) depends on numpy (>=1.15.0)
INFO: fact: spacy (3.3.0) depends on packaging (>=20.0)
INFO: fact: spacy (3.3.0) depends on pathy (>=0.3.5)
INFO: fact: spacy (3.3.0) depends on preshed (<3.1.0,>=3.0.2)
INFO: fact: spacy (3.3.0) depends on pydantic (>=1.7.4,<1.8 || >1.8,<1.8.1 || >1.8.1,<1.9.0)
INFO: fact: spacy (3.3.0) depends on requests (<3.0.0,>=2.13.0)
INFO: fact: spacy (3.3.0) depends on setuptools (*)
INFO: fact: spacy (3.3.0) depends on spacy-legacy (<3.1.0,>=3.0.9)
INFO: fact: spacy (3.3.0) depends on spacy-loggers (>=1.0.0,<2.0.0)
INFO: fact: spacy (3.3.0) depends on srsly (<3.0.0,>=2.4.3)
INFO: fact: spacy (3.3.0) depends on thinc (>=8.0.14,<8.1.0)
INFO: fact: spacy (3.3.0) depends on tqdm (<5.0.0,>=4.38.0)
INFO: fact: spacy (3.3.0) depends on typer (<0.5.0,>=0.3.0)
INFO: fact: spacy (3.3.0) depends on wasabi (<1.1.0,>=0.9.1)
INFO: derived: not spacy (3.3.0)
INFO: discovering spacy==3.2.5
INFO: fact: spacy (3.2.5) depends on blis (<0.8.0,>=0.4.0)
INFO: fact: spacy (3.2.5) depends on catalogue (>=2.0.6,<2.1.0)
INFO: fact: spacy (3.2.5) depends on cymem (>=2.0.2,<2.1.0)
INFO: fact: spacy (3.2.5) depends on jinja2 (*)
INFO: fact: spacy (3.2.5) depends on langcodes (<4.0.0,>=3.2.0)
INFO: fact: spacy (3.2.5) depends on murmurhash (<1.1.0,>=0.28.0)
INFO: fact: spacy (3.2.5) depends on numpy (>=1.15.0)
INFO: fact: spacy (3.2.5) depends on packaging (>=20.0)
INFO: fact: spacy (3.2.5) depends on pathy (>=0.3.5)
INFO: fact: spacy (3.2.5) depends on preshed (<3.1.0,>=3.0.2)
INFO: fact: spacy (3.2.5) depends on pydantic (>=1.7.4,<1.8 || >1.8,<1.8.1 || >1.8.1,<1.9.0)
INFO: fact: spacy (3.2.5) depends on requests (<3.0.0,>=2.13.0)
INFO: fact: spacy (3.2.5) depends on setuptools (*)
INFO: fact: spacy (3.2.5) depends on smart-open (<7.0.0,>=5.2.1)
INFO: fact: spacy (3.2.5) depends on spacy-legacy (<3.1.0,>=3.0.8)
INFO: fact: spacy (3.2.5) depends on spacy-loggers (>=1.0.0,<2.0.0)
INFO: fact: spacy (3.2.5) depends on srsly (<3.0.0,>=2.4.1)
INFO: fact: spacy (3.2.5) depends on thinc (>=8.0.12,<8.1.0)
INFO: fact: spacy (3.2.5) depends on tqdm (<5.0.0,>=4.38.0)
INFO: fact: spacy (3.2.5) depends on typer (<0.5.0,>=0.3.0)
INFO: fact: spacy (3.2.5) depends on wasabi (<1.1.0,>=0.8.1)
INFO: derived: not spacy (3.2.5)
INFO: discovering spacy==3.2.4
INFO: fact: spacy (3.2.4) depends on blis (<0.8.0,>=0.4.0)
INFO: fact: spacy (3.2.4) depends on catalogue (>=2.0.6,<2.1.0)
INFO: fact: spacy (3.2.4) depends on click (<8.1.0)
INFO: fact: spacy (3.2.4) depends on cymem (>=2.0.2,<2.1.0)
INFO: fact: spacy (3.2.4) depends on jinja2 (*)
INFO: fact: spacy (3.2.4) depends on langcodes (<4.0.0,>=3.2.0)
INFO: fact: spacy (3.2.4) depends on murmurhash (<1.1.0,>=0.28.0)
INFO: fact: spacy (3.2.4) depends on numpy (>=1.15.0)
INFO: fact: spacy (3.2.4) depends on packaging (>=20.0)
INFO: fact: spacy (3.2.4) depends on pathy (>=0.3.5)
INFO: fact: spacy (3.2.4) depends on preshed (<3.1.0,>=3.0.2)
INFO: fact: spacy (3.2.4) depends on pydantic (>=1.7.4,<1.8 || >1.8,<1.8.1 || >1.8.1,<1.9.0)
INFO: fact: spacy (3.2.4) depends on requests (<3.0.0,>=2.13.0)
INFO: fact: spacy (3.2.4) depends on setuptools (*)
INFO: fact: spacy (3.2.4) depends on spacy-legacy (<3.1.0,>=3.0.8)
INFO: fact: spacy (3.2.4) depends on spacy-loggers (>=1.0.0,<2.0.0)
INFO: fact: spacy (3.2.4) depends on srsly (<3.0.0,>=2.4.1)
INFO: fact: spacy (3.2.4) depends on thinc (>=8.0.12,<8.1.0)
INFO: fact: spacy (3.2.4) depends on tqdm (<5.0.0,>=4.38.0)
INFO: fact: spacy (3.2.4) depends on typer (<0.5.0,>=0.3.0)
INFO: fact: spacy (3.2.4) depends on wasabi (<1.1.0,>=0.8.1)
INFO: derived: not spacy (3.2.4)
INFO: discovering spacy==3.2.3
INFO: fact: spacy (3.2.3) depends on blis (<0.8.0,>=0.4.0)
INFO: fact: spacy (3.2.3) depends on catalogue (>=2.0.6,<2.1.0)
INFO: fact: spacy (3.2.3) depends on cymem (>=2.0.2,<2.1.0)
INFO: fact: spacy (3.2.3) depends on jinja2 (*)
INFO: fact: spacy (3.2.3) depends on langcodes (<4.0.0,>=3.2.0)
INFO: fact: spacy (3.2.3) depends on murmurhash (<1.1.0,>=0.28.0)
INFO: fact: spacy (3.2.3) depends on numpy (>=1.15.0)
INFO: fact: spacy (3.2.3) depends on packaging (>=20.0)
INFO: fact: spacy (3.2.3) depends on pathy (>=0.3.5)
INFO: fact: spacy (3.2.3) depends on preshed (<3.1.0,>=3.0.2)
INFO: fact: spacy (3.2.3) depends on pydantic (>=1.7.4,<1.8 || >1.8,<1.8.1 || >1.8.1,<1.9.0)
INFO: fact: spacy (3.2.3) depends on requests (<3.0.0,>=2.13.0)
INFO: fact: spacy (3.2.3) depends on setuptools (*)
INFO: fact: spacy (3.2.3) depends on spacy-legacy (<3.1.0,>=3.0.8)
INFO: fact: spacy (3.2.3) depends on spacy-loggers (>=1.0.0,<2.0.0)
INFO: fact: spacy (3.2.3) depends on srsly (<3.0.0,>=2.4.1)
INFO: fact: spacy (3.2.3) depends on thinc (>=8.0.12,<8.1.0)
INFO: fact: spacy (3.2.3) depends on tqdm (<5.0.0,>=4.38.0)
INFO: fact: spacy (3.2.3) depends on typer (<0.5.0,>=0.3.0)
INFO: fact: spacy (3.2.3) depends on wasabi (<1.1.0,>=0.8.1)
INFO: derived: not spacy (3.2.3)
INFO: discovering spacy==3.2.2
INFO: fact: spacy (3.2.2) depends on blis (<0.8.0,>=0.4.0)
INFO: fact: spacy (3.2.2) depends on catalogue (>=2.0.6,<2.1.0)
INFO: fact: spacy (3.2.2) depends on cymem (>=2.0.2,<2.1.0)
INFO: fact: spacy (3.2.2) depends on jinja2 (*)
INFO: fact: spacy (3.2.2) depends on langcodes (<4.0.0,>=3.2.0)
INFO: fact: spacy (3.2.2) depends on murmurhash (<1.1.0,>=0.28.0)
INFO: fact: spacy (3.2.2) depends on numpy (>=1.15.0)
INFO: fact: spacy (3.2.2) depends on packaging (>=20.0)
INFO: fact: spacy (3.2.2) depends on pathy (>=0.3.5)
INFO: fact: spacy (3.2.2) depends on preshed (<3.1.0,>=3.0.2)
INFO: fact: spacy (3.2.2) depends on pydantic (>=1.7.4,<1.8 || >1.8,<1.8.1 || >1.8.1,<1.9.0)
INFO: fact: spacy (3.2.2) depends on requests (<3.0.0,>=2.13.0)
INFO: fact: spacy (3.2.2) depends on setuptools (*)
INFO: fact: spacy (3.2.2) depends on spacy-legacy (<3.1.0,>=3.0.8)
INFO: fact: spacy (3.2.2) depends on spacy-loggers (>=1.0.0,<2.0.0)
INFO: fact: spacy (3.2.2) depends on srsly (<3.0.0,>=2.4.1)
INFO: fact: spacy (3.2.2) depends on thinc (>=8.0.12,<8.1.0)
INFO: fact: spacy (3.2.2) depends on tqdm (<5.0.0,>=4.38.0)
INFO: fact: spacy (3.2.2) depends on typer (<0.5.0,>=0.3.0)
INFO: fact: spacy (3.2.2) depends on wasabi (<1.1.0,>=0.8.1)
INFO: derived: not spacy (3.2.2)
INFO: discovering spacy==3.2.1
...
<output here>
```

</p></details>

#### What you expected to happen

Resolved in < 1 min (pip-compile takes 16 secs).

**Comment by ddelange** [comment]
Hi @tekumara, it looks like you're talking about spending a long time on `spacy`. In your case, additionally providing e.g. `spacy~=3.3` to pipgrip will probably speed things up for you. I'll close this as a duplicate of

**Comment by tekumara** [comment]
Thanks for that... the interesting thing is that `pip-compile --resolver=backtracking` collects spacy 3.3.2 and doesn't backtrack further, so it avoids the need to build the older versions.

**Comment by ddelange** [comment]
So it only looks at one older version only. That sounds like an early backtrack. Interesting 🤔

**Comment by tekumara** [comment]
It's possible pip-compile is wrong tho 😅 I notice poetry selects spacy 2.3.9

**Comment by ddelange** [comment]
Providing lower pins for both your libs [solves](https://colab.research.google.com/drive/1R1AoG3uUlpjUw1M3an_IvyZOP1qNrnBp#scrollTo=Kd6ugjF5wTIm) in under 5 mins, re-run with pip cache in about 2.5 mins.

```
pipgrip -vv --lock --sort --tree allennlp~=2.10 spacy~=3.3
```

**Comment by ddelange** [comment]
fixed link permissions ^
